File: week-1-c/credit.c
#include <cs50.h>
#include <stdbool.h>
#include <stdio.h>

void findCreditType(long creditNo);
int findLength(long creditNo);
long findSubPart(int n, long creditNo);
bool isCheckSumValid(long creditNo);
int findSumOfDigits(int twiceOfCurrentLastDigit);

const string invalid = "INVALID\n";

int main(void)
{
    // Taking credit number as input from the user
    long creditNo = get_long("Enter the credit card number: ");

    // printf("length: %i\n", findLength(creditNo));
    // printf("From calculateCheckSum: %s\n", calculateCheckSum(creditNo));

    // Calculating the checkSum of the credit number and validating it
    bool checkSumValidOrInvalid = isCheckSumValid(creditNo);

    if (checkSumValidOrInvalid)
    {
        // if checkSum is valid, find the credit type
        findCreditType(creditNo);
    }
    else
    {
        // if checkSum is invalid, print INVALID
        printf("%s", invalid);
    }
}

bool isCheckSumValid(long creditNo)
{
    int checkSum = 0;
    int currentLastDigit;
    bool result = false;

    for (long i = creditNo, counter = 1; i > 0; i /= 10)
    {
        currentLastDigit = i % 10;
        if (counter % 2 == 0)
        {
            int twiceOfCurrentLastDigit = currentLastDigit * 2;

            if (twiceOfCurrentLastDigit > 9)
            {
                twiceOfCurrentLastDigit = findSumOfDigits(twiceOfCurrentLastDigit);
            }

            checkSum += twiceOfCurrentLastDigit; // Add sum of the digits of the product
        }
        else
        {
            checkSum += currentLastDigit;
        }
        counter++;
    }

    // printf("checkSum: %i\n", checkSum);
    if (checkSum % 10 == 0)
    {
        result = true;
    }

    return result;
}

int findSumOfDigits(int twiceOfCurrentLastDigit)
{
    int sumOfDigits = 0;

    for (int i = twiceOfCurrentLastDigit; i > 0; i /= 10)
    {
        sumOfDigits += (i % 10);
    }

    return sumOfDigits;
}

// American Express - 15 digits, starts with 34 or 37
// MasterCard - 16 digits, starts with 51, 52, 53, 54, 55
// Visa - 13 or 16 digits, starts with 4
void findCreditType(long creditNo)
{
    int length = findLength(creditNo);

    if (length == 15 && (findSubPart(2, creditNo) == 34 || findSubPart(2, creditNo) == 37))
    {
        printf("AMEX\n");
    }
    else if (length == 16 && (findSubPart(2, creditNo) == 51 || findSubPart(2, creditNo) == 52 ||
                              findSubPart(2, creditNo) == 53 || findSubPart(2, creditNo) == 54 ||
                              findSubPart(2, creditNo) == 55))
    {
        printf("MASTERCARD\n");
    }
    else if ((length == 13 || length == 16) && findSubPart(1, creditNo) == 4)
    {
        printf("VISA\n");
    }
    else
    {
        printf("%s", invalid);
    }
}

long findSubPart(int n, long creditNo)
{
    long subPart = 0;
    for (long i = creditNo; i > 0; i /= 10)
    {
        if (findLength(i) == n)
        {
            subPart = i;
            break;
        }
    }
    // printf("subPart: %li\n", subPart);
    return subPart;
}

int findLength(long creditNo)
{
    int length = 0;
    for (long i = creditNo; i > 0; i /= 10)
    {
        length++;
    }
    return length;
}

File: week-6-python/credit.py
from cs50 import get_string

# Find the checkSum of the credit number provided


def is_checksum_valid(creditNo):
    checkSum = 0
    counter = 0

    # Traversing from the last index of the credit no.
    for i in range(len(creditNo), -1, -1):

        # Skipping the length because it will cause Index Out Of Bound Exception
        if i != len(creditNo):
            # print("counter", counter, i)

            # When there is a even value of counter
            if counter % 2 == 0:
                # print(int(creditNo[i]))

                # doubling the value
                twiceOfCurrentLastDigit = int(creditNo[i]) * 2
                # print(twiceOfCurrentLastDigit)

                # if the doubled value is >9, find the sum of the digits and add that to checksum
                if (twiceOfCurrentLastDigit > 9):
                    twiceOfCurrentLastDigit = findSumOfDigits(twiceOfCurrentLastDigit)
                    # print("After sum of digits", twiceOfCurrentLastDigit)

                # print(checkSum)
                checkSum = checkSum + twiceOfCurrentLastDigit
                # print("Updated:", checkSum)

            # if counter value is odd value, just add the value to the checksum
            else:
                # print(checkSum)
                # print(int(creditNo[i]))
                checkSum = checkSum + int(creditNo[i])
                # print("Updated", checkSum)

        # increment the counter to track the position
        counter = counter + 1

    # print("Finally", checkSum)

    if checkSum % 10 == 0:
        return True

    return False

# Find the credit card type


def findCreditCardType(creditNo):
    if len(creditNo) == 15 and (creditNo[0:2] in ["34", "37"]):
        print("AMEX")
    elif len(creditNo) == 16 and (creditNo[0:2] in ["51", "52", "53", "54", "55"]):
        print("MASTERCARD")
    elif len(creditNo) in [13, 16] and creditNo[0] == "4":
        print("VISA")
    else:
        print("INVALID")


def main():

    while (True):
        creditNo = get_string("Enter credit card number: ")
        if (creditNo.isnumeric()):
            break

    isCheckSumValid = is_checksum_valid(creditNo)

    if (isCheckSumValid):
        findCreditCardType(creditNo)
    else:
        print("INVALID")


# Find the sum of the digits of a number
def findSumOfDigits(twiceOfCurrentLastDigit):
    sumOfDigits = 0

    while (twiceOfCurrentLastDigit > 0):
        sumOfDigits = sumOfDigits + round(twiceOfCurrentLastDigit % 10)
        twiceOfCurrentLastDigit = twiceOfCurrentLastDigit // 10

    return sumOfDigits


main()
